From c6ecc80fbe4ff28fac6ffc7de5e8263d5dc25976 Mon Sep 17 00:00:00 2001
From: Chunming Zhou <David1.Zhou@amd.com>
Date: Fri, 10 Feb 2017 16:46:09 +0800
Subject: [PATCH 3075/3082] drm/amdgpu: fix get_file only when multiple
 sem_export

file put will be hanled by close(fd).

Change-Id: I82b4693d39216674b94df05760b79d3d5964985e
Signed-off-by: Chunming Zhou <David1.Zhou@amd.com>
Reviewed-by: David Mao <David.Mao@amd.com>
---
 drivers/gpu/drm/amd/amdgpu/amdgpu_sem.c | 6 ++----
 1 file changed, 2 insertions(+), 4 deletions(-)

diff --git a/drivers/gpu/drm/amd/amdgpu/amdgpu_sem.c b/drivers/gpu/drm/amd/amdgpu/amdgpu_sem.c
index 57b6dd8..b68abf17 100644
--- a/drivers/gpu/drm/amd/amdgpu/amdgpu_sem.c
+++ b/drivers/gpu/drm/amd/amdgpu/amdgpu_sem.c
@@ -45,9 +45,6 @@ static void amdgpu_sem_core_free(struct kref *kref)
 	struct amdgpu_sem_core *core = container_of(
 		kref, struct amdgpu_sem_core, kref);
 
-	if (core->file)
-		fput(core->file);
-
 	fence_put(core->fence);
 	mutex_destroy(&core->lock);
 	kfree(core);
@@ -278,10 +275,11 @@ static int amdgpu_sem_export(struct amdgpu_fpriv *fpriv,
 			ret = -ENOMEM;
 			goto err_put_sem;
 		}
+	} else {
+		get_file(core->file);
 	}
 	mutex_unlock(&core->lock);
 
-	get_file(core->file);
 	kref_get(&core->kref);
 
 	ret = get_unused_fd_flags(O_CLOEXEC);
-- 
2.7.4

